### DISCUSSION TITLE
v6: ESBuild issue, The argument 'filename' must be a file URL object, file URL string, or absolute path string. Received undefined",

### DISCUSSION THREAD

**Discussion by paul-uz** [openpgpjs/openpgpjs#1818] [Q&A]
Using the following esbuild setup to compile TS files into single JS files, we see the error below:

esbuild
```
const output = await esbuild.build({
      entryPoints: [entryPoint],
      platform: 'node',
      target: 'esnext',
      bundle: true,
      sourcemap: true,
      outdir: `.dist/lambda/${folderName}/`,
      logLevel: 'silent', // Suppress default logging
      loader: {
        '.node': 'file', // Specify that .node files should be loaded as files
      },
      external: [
        '@aws-sdk',
      ],
      plugins: [
        requireResolvePlugin(),
      ],
    });
```

error
```
{
  "errorType": "TypeError",
  "errorMessage": "The argument 'filename' must be a file URL object, file URL string, or absolute path string. Received undefined",
  "trace": [
    "TypeError [ERR_INVALID_ARG_VALUE]: The argument 'filename' must be a file URL object, file URL string, or absolute path string. Received undefined",
    "    at createRequire (node:internal/modules/cjs/loader:1760:11)",
    "    at Object.<anonymous> (/var/task/index.js:12229:48)",
    "    at Module._compile (node:internal/modules/cjs/loader:1546:14)",
    "    at Object..js (node:internal/modules/cjs/loader:1689:10)",
    "    at Module.load (node:internal/modules/cjs/loader:1318:32)",
    "    at Function._load (node:internal/modules/cjs/loader:1128:12)",
    "    at TracingChannel.traceSync (node:diagnostics_channel:315:14)",
    "    at wrapModuleLoad (node:internal/modules/cjs/loader:218:24)",
    "    at Module.require (node:internal/modules/cjs/loader:1340:12)",
    "    at require (node:internal/modules/helpers:141:16)"
  ]
}
```

It looks like the offending line is

`nodeRequire: (0, import_module.createRequire)(import_meta.url),`

I've searched and searched and found semi relevant topics, but cannot get any "solutions" to work.

**Comment by twiss** [comment]
Hi :wave: Which version and build are you using?
From the code it looks like you're using the Node.js build. Before we debug further, could you try using the browser build (e.g. by importing `openpgp/dist/openpgp.mjs` directly), and see if that solves your issue?

**Comment by paul-uz** [comment]
@twiss like this? `import * as openpgp from '../../node_modules/openpgp/dist/openpgp.mjs';`

EDIT - this seems to work. Is there a better way to do the import?

**Comment by fixparser** [comment]
+1 on the same issue.
I previously used version `5.11.2` and updated to `6.0.1` and ran into these issues:
* The above ESBuild issue
* Now only able to run a local dev server with HTTPS, otherwise I get a "Web Crypto is not available" error
* OpenPGP seems to be published differently now (as MJS), causing issues with Jest (for now I just mock OpenPGP to get around the issue)

Thanks for the great work!

I'm using this library in my `fixparser` NPM package for license validation purposes. https://fixparser.dev

**Comment by twiss** [comment]
You can indicate to ESBuild that you want the browser build using `platform: 'browser'`, though that might affect other things as well. Also for OpenPGP.js, if you are in fact running on Node.js, it might hurt performance.

If you want the Node.js build, you might need [this issue](https://github.com/evanw/esbuild/issues/3099) to be fixed, or shim `import.meta.url` yourself somehow.

**Comment by fixparser** [comment]
> You can indicate to ESBuild that you want the browser build using `platform: 'browser'`, though that might affect other things as well. Also for OpenPGP.js, if you are in fact running on Node.js, it might hurt performance.
> 
> If you want the Node.js build, you might need [this issue](https://github.com/evanw/esbuild/issues/3099) to be fixed, or shim `import.meta.url` yourself somehow.

Yes, the `fixparser` library uses both `platform: browser` and `platform: node` to build for multiple environments.
Thanks for sharing the ESBuild issue, will follow up on that! Cheers

**Comment by Maksim-Maltsev** [comment]
+1 - same happens in AWS lambda environment, in 5.1.2 were no issues.

**Comment by matthew-gladman-oua** [comment]
We also ran into this issue today, building via esbuild for node only as well

**Comment by larabr** [comment]
With v6 we introduced some changes in the bundling, as mentioned the problem here is with ESBuild.

You'll get this error with `platform=node` if you `import` openpgp (as oppose to `require` it) and set ESBuild to output a non-ESM format (default is `format=cjs`). With this setup ESBuild will try and transform our Node .mjs bundle and choke on `import.meta`.

For more context about the issue and possible manual solutions (i.e. use `format: 'esm'` or replace `'import.meta.url'` using a custom plugin), see https://github.com/evanw/esbuild/issues/1492 .

FYI ESBuild might be printing out some relevant warnings about `import.meta` failing to transform based on the config, e.g.:
```
▲ [WARNING] "import.meta" is not available with the "cjs" output format and will be empty [empty-import-meta]

    dist/node/openpgp.min.mjs:2:13261:
      2 │ ...){return"string"==typeof e||e instanceof String},nodeRequire:t(import.meta.url),isArray:function(e){return e instanceof Array},isUint8Ar...
        ╵                                                                   ~~~~~~~~~~~

  You need to set the output format to "esm" for "import.meta" to work correctly.
```

**Comment by 0cnLaroche** [comment]
Ran into the same issue with  `6.0.1`, reverted to `5.11.2` and it's working now

**Comment by eduardoprentcomau** [comment]
Worked for me using Serverless Framework by including:

```js
external: [
  'openpgp',
],
```

**Comment by EhabY** [comment]
I added the following settings and it seems to work and is being bundled correctly now (since I'm using cjs anyway):

```js
format: "cjs",
mainFields: ["module", "main"],
// Force openpgp to use CJS. The ESM version uses import.meta.url which is
// undefined when bundled to CJS, causing runtime errors.
alias: {
	openpgp: "./node_modules/openpgp/dist/node/openpgp.min.cjs",
},
```
